src/include/sof/audio/module_adapter/module/module_interface.h

```diff
@@ -81,15 +83,53 @@ struct module_interface {
 	 * component preparation in .prepare()
 	 */
 	int (*prepare)(struct processing_module *mod);
+
 	/**
-	 * Module specific processing procedure, called as part of module_adapter
-	 * component copy in .copy(). This procedure is responsible to consume
+	 * Module specific processing procedure
+	 * This procedure is responsible to consume
 	 * samples provided by the module_adapter and produce/output the processed
 	 * ones back to module_adapter.
+	 *
+	 * there are 3 versions of the procedure, the difference is the format of
+	 * input/output data
+	 *
+	 * the module MUST implement one and ONLY one of them
```

Comment:
use a `union`?

Reply:
We talked about it already ;) and it has been rejected

3 function pointers are to show
1) there are 3 APIs here (raw/stream/snk_src)
2) make module_adapter aware what API the module uses

alternatively I need to introduce an enum, like simple_copy,  and make the module set it.